addon/record-data.js

```diff
@@ -375,170 +485,366 @@ export default class FragmentRecordData extends RecordData {
+      if (data == null) {
+        continue;
       }
-      delete attributes[key];
-    });
-    return super.didCommit(data);
+      behavior.commitWasRejected(data);
+    }
+    Object.assign(this._fragments, this._inFlightFragments);
+    this._inFlightFragments = null;
+    super.commitWasRejected(identifier, errors);
+  }
+
+  rollbackAttributes() {
+    let dirtyFragmentKeys;
+    if (this.hasChangedFragments()) {
+      dirtyFragmentKeys = Object.keys(this._fragments);
+      dirtyFragmentKeys.forEach((key) => {
+        this.rollbackFragment(key);
+      });
+      this._fragments = null;
+    }
+    const dirtyAttributeKeys = super.rollbackAttributes();
+    this.notifyStateChange();
+    this.fragmentDidReset();
+    return mergeArrays(dirtyAttributeKeys, dirtyFragmentKeys);
+  }
+
+  rollbackFragment(key) {
+    const behavior = this._fragmentBehavior[key];
+    assert(`Attribute '${key}' for model '${this.modelName}' must be a fragment`, behavior != null);
+    if (!this.isFragmentDirty(key)) {
+      return;
+    }
+    delete this._fragments[key];
+    const fragment = this._fragmentData[key];
+    if (fragment == null) {
+      return;
+    }
+    behavior.rollback(fragment);
+    this._fragmentArrayCache[key]?.notify();
+
+    if (!this.hasChangedAttributes()) {
+      this.notifyStateChange(key);
+      this.fragmentDidReset();
+    }
+  }
+
+  reset() {
+    super.reset();
+    this.__fragments = null;
+    this.__inFlightFragments = null;
+    this.__fragmentData = null;
+    this.__fragmentArrayCache = null;
+    this._fragmentOwner = null;
   }
 
-  commitWasRejected() {
-    let key, fragment;
-    for (key in this.inFlightFragments) {
-      fragment = this.inFlightFragments[key];
-      delete this.inFlightFragments[key];
-      if (this.fragments[key] === undefined) {
-        this.fragments[key] = fragment;
+  unloadRecord() {
+    for (const [key, behavior] of Object.entries(this._fragmentBehavior)) {
+      const fragment = this._fragments[key];
+      if (fragment != null) {
+        behavior.unload(fragment);
       }
-      if (fragment) {
-        // TODO this bad, we should keep track of fragment record datas ourself
-        if (fragment.content) {
-          fragment.content.forEach(frag => {
-            // check to see if the array is a non-fragment array, if so, don't call
-            // method on _internalModel.recordData
-            // TODO: We need to add inFlight details to statefulArray, then change this.
-            if (frag._internalModel) {
-              frag._internalModel._recordData.commitWasRejected();
-            }
-          });
-        } else {
-          fragment._internalModel._recordData.commitWasRejected();
-        }
+      const inFlight = this._inFlightFragments[key];
+      if (inFlight != null) {
+        behavior.unload(inFlight);
       }
+      const fragmentData = this._fragmentData[key];
+      if (fragmentData != null) {
+        behavior.unload(fragmentData);
```

Comment:
Testing this out against my test suite, I'm getting a bunch of errors from here.
```
Assertion Failed: You can only unload a record which is not inFlight.
at Object.assert (http://localhost:7357/app/assets/vendor.js:32841:15)
    at Object.assertAgainstUnloadRecord [as unloadRecord] (http://localhost:7357/app/assets/vendor.js:117846:12)
    at InternalModel.send (http://localhost:7357/app/assets/vendor.js:118720:32)
    at InternalModel.unloadRecord (http://localhost:7357/app/assets/vendor.js:118405:12)
    at FragmentRecordData._fragmentUnloadRecord (http://localhost:7357/app/assets/vendor.js:145645:30)
    at FragmentBehavior.unload (http://localhost:7357/app/assets/vendor.js:145096:16)
    at FragmentRecordData.unloadRecord (http://localhost:7357/app/assets/vendor.js:145565:20)
```